File: articles/dynamic-complexity-and-policy-resistance/r/scenario_comparison.R
# Scenario comparison for dynamic complexity.
path <- file.path("data", "raw", "synthetic_scenarios.csv")
if (file.exists(path)) {
  scenarios <- read.csv(path)
  scenarios$resistance_index <- round(scenarios$compensation_strength + scenarios$boundary_cost_weight / 2, 3)
  print(scenarios[order(-scenarios$resistance_index), ])
}

File: _archive/pre-professional-script-upgrade/20260531-194442/reinforcing-and-balancing-dynamics/python/scenario_comparison.py
"""Scenario comparison for reinforcing and balancing dynamics."""

from __future__ import annotations

import csv
from pathlib import Path

DATA = Path(__file__).resolve().parents[1] / "data" / "synthetic_scenarios.csv"


def run_scenario(row: dict[str, str], periods: int = 20) -> float:
    value = 10.0
    rate = float(row["growth_rate"])
    correction = float(row["correction_strength"])
    capacity = float(row["carrying_capacity"])
    for _ in range(periods):
        reinforcing = rate * value
        balancing = correction * (value / capacity) * value
        value = value + reinforcing - balancing
    return round(value, 4)


def main() -> None:
    with DATA.open(newline="", encoding="utf-8") as handle:
        rows = list(csv.DictReader(handle))
    print("scenario_id,scenario_name,final_value")
    for row in rows:
        print(f"{row['scenario_id']},{row['scenario_name']},{run_scenario(row)}")


if __name__ == "__main__":
    main()
